File: Linear Data Structures - Stacks and Queues/homework/linked_stack.py
class Node:
    def __init__(self, value, bottom_node: 'Node'=None):
        self.value = value
        self.__bottom_node = bottom_node

    @property
    def bottom_node(self):
        return self.__bottom_node

    @bottom_node.setter
    def bottom_node(self, node):
        self.__bottom_node = node


class LinkedStack:
    def __init__(self):
        self.first_node = None
        self.count = 0

    def __iter__(self):
        node = self.first_node
        while node:
            yield node.value
            node = node.bottom_node

    def push(self, element):
        if self.count == 0:
            self.first_node = Node(value=element, bottom_node=None)
        else:
            new_node = Node(value=element, bottom_node=self.first_node)
            self.first_node = new_node
        self.count += 1

    def pop(self):
        if self.count == 0:
            raise Exception('Stack is empty!')
        node_value = self.first_node.value
        self.first_node = self.first_node.bottom_node
        self.count -= 1
        return node_value

File: LinkedList/LinkedStack.java
public class LinkedStack<E> {

    private Node<E> firstNode;
    private int size;


    public int size() {
        return this.size;
    }

    private void setSize(int size) {
        this.size = size;
    }

    public void push(E element) {
        Node<E> newNode = new Node<E>(element, this.firstNode);
        this.firstNode = newNode;
        this.size++;
    }

    public E pop() {
        E el = this.firstNode.value;
        this.firstNode = this.firstNode.nextNode;
        this.size--;
        return el;
    }

    public E[] toArray() {
        E tkn[] = (E[])new Object[this.size];
        if (firstNode == null) {
            return tkn;
        }
        int idx = 0;
        while (firstNode.nextNode != null) {
            tkn[idx] = firstNode.value;
            firstNode = firstNode.nextNode;
            idx++;
        }
        tkn[idx] = firstNode.value;
        return tkn;
    }

    private class Node<E> {

        private E value;
        private Node<E> nextNode;

        public Node(E value) {
            this.value = value;
            this.nextNode = null;
        }

        public Node(E value, Node<E> nextNode) {
            this.value = value;
            this.nextNode = nextNode;
        }

        public Node<E> getNextNode() {
            return this.nextNode;
        }

        public void setNextNode(Node<E> nextNode) {
            this.nextNode = nextNode;
        }
    }
}